wooting-analog-sdk/src/plugin.rs

```diff
@@ -559,3 +628,242 @@ impl Plugin for WootingPlugin {
+    use ffi_support::FfiStr;
+    use libloading::{Library, Symbol};
+    use log::*;
+    use log::{error, info};
+    use std::collections::HashMap;
+    use std::os::raw::{c_float, c_int, c_uint, c_ushort, c_void};
+
+    use crate::{
+        DeviceEventType, DeviceID, DeviceInfo, DeviceInfo_FFI, Plugin, SDKResult,
+        WootingAnalogResult,
+    };
+
+    macro_rules! lib_wrap {
+    //(@as_item $i:item) => {$i};
+
+    (
+        $(
+            fn $fn_names:ident($($fn_arg_names:ident: $fn_arg_tys:ty),*) $(-> $fn_ret_tys:ty)*;
+        )*
+    ) => {
+        $(
+            //lib_wrap! {
+            //    @as_item
+                #[no_mangle]
+                fn $fn_names(&mut self, $($fn_arg_names: $fn_arg_tys),*) $(-> $fn_ret_tys)* {
+                    unsafe {
+                        type FnPtr = unsafe fn($($fn_arg_tys),*) $(-> $fn_ret_tys)*;
+                        //TODO: Retain the obtained function pointer between calls
+                        let func :  Option<Symbol<FnPtr>>  = self.lib.get(stringify!($fn_names).as_bytes()).map_err(|e| {
+                                    error!("{}", e);
+                                }).ok();
+                        match func {
+                            Some(f) => f($($fn_arg_names),*).into(),
+                            _ => Default::default()
+
+                        }
+                    }
+                }
+            //}
+        )*
+    };
+}
+
+    macro_rules! lib_wrap_option {
+    //(@as_item $i:item) => {$i};
+
+    (
+        $(
+            fn $fn_names:ident($($fn_arg_names:ident: $fn_arg_tys:ty),*) $(-> $fn_ret_tys:ty)*;
+        )*
+    ) => {
+        $(
+            //lib_wrap! {
+            //    @as_item
+                #[no_mangle]
+                fn $fn_names(&mut self, $($fn_arg_names: $fn_arg_tys),*) $(-> SDKResult<$fn_ret_tys>)* {
+                    unsafe {
+                        type FnPtr = unsafe fn($($fn_arg_tys),*) $(-> $fn_ret_tys)*;
+                        let func :Option<Symbol<FnPtr>>  = self.lib.get(stringify!($fn_names).as_bytes()).map_err(|e| {
+                                    error!("{}", e);
+                                }).ok();
+                        match func {
+                            Some(f) => f($($fn_arg_names),*).into(),
+                            _ => Err(WootingAnalogResult::FunctionNotFound).into()
+
+                        }
+                    }
+                }
+            //}
+        )*
+    };
+}
+
+    const CPLUGIN_ABI_VERSION: u32 = 1;
+
+    pub struct CPlugin {
+        lib: Library,
+        cb_data_ptr: Option<*mut Box<dyn Fn(DeviceEventType, &DeviceInfo) + Send>>,
+        //funcs: HashMap<&'static str, Option<Symbol>>
+    }
+
+    impl CPlugin {
+        pub fn new(lib: Library) -> SDKResult<CPlugin> {
+            unsafe {
+                if let Some(ver) = lib.get::<*mut u32>(b"ANALOG_SDK_PLUGIN_ABI_VERSION").ok() {
+                    let v = **ver;
```

Comment:
All this cplugin stuff needs to be in a different file than the device handling implementation